runtime/src/utils/buffer/write.rs

```diff
@@ -176,153 +217,99 @@ impl<B: Blob> Blob for Write<B> {
+            .ok_or(Error::OffsetOverflow)?;
 
-        // Case 2: Write can be merged into existing buffer.
-        //
-        // This handles overwrites and extensions within the buffer's current capacity,
-        // including writes that create gaps (filled with zeros) in the buffer.
-        let can_merge_into_buffer = offset >= buffer_start
-            && (offset - buffer_start) + data_len as u64 <= inner.capacity as u64;
-        if can_merge_into_buffer {
-            let buffer_offset = (offset - buffer_start) as usize;
-            let required_len = buffer_offset + data_len;
-
-            // Expand buffer if necessary (fills with zeros)
-            if required_len > inner.buffer.len() {
-                inner.buffer.resize(required_len, 0);
-            }
+        // Acquire a write lock on the buffer.
+        let mut buffer = self.buffer.write().await;
 
-            // Copy data into buffer
-            inner.buffer[buffer_offset..required_len].copy_from_slice(data);
+        // Write falls entirely within the buffer's current range and can be merged.
+        if buffer.merge(buf.as_ref(), offset) {
             return Ok(());
         }
 
-        // Case 3: Write cannot be merged - flush buffer and write directly.
-        //
-        // This includes: writes before the buffer, writes that would exceed capacity,
-        // or non-contiguous writes that can't be merged.
-        if !inner.buffer.is_empty() {
-            inner.flush().await?;
-        }
-        inner.blob.write_at(buf, offset).await?;
-
-        // Update position to maintain "buffer at tip" invariant.
-        //
-        // Position should advance to the end of this write if it extends the logical blob
-        let write_end = offset + data_len as u64;
-        if write_end > inner.position {
-            inner.position = write_end;
+        // Write cannot be merged, so flush the buffer if the range overlaps, and check if merge is
+        // possible after.
+        if buffer.offset < end_offset {
+            if let Some((old_buf, buffer_offset)) = buffer.take_buf() {
+                self.blob.write_at(old_buf, buffer_offset).await?;
+                if buffer.merge(buf.as_ref(), offset) {
+                    return Ok(());
+                }
+            }
         }
 
+        // Write could not be merged (exceeds buffer capacity or outside its range), so write
+        // directly. Note that we end up writing an intersecting range twice: once when the buffer
+        // is flushed above, then again when we write the `buf` below. Removing this inefficiency
+        // may not be worth the additional complexity.
+        self.blob.write_at(buf, offset).await?;
+
+        // Maintain the "buffer at tip" invariant by advancing offset to the end of this write if it
+        // extended the underlying blob.
+        buffer.offset = buffer.offset.max(end_offset);
+
         Ok(())
     }
 
     async fn truncate(&self, len: u64) -> Result<(), Error> {
```

Comment:
We'll need to adjust this to `resize`: https://github.com/commonwarexyz/monorepo/pull/1167